compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/writer/BIRTypeWriter.java

```diff
@@ -577,4 +589,104 @@ private void writeTypeInclusions(List<BType> inclusions) {
+    }
+
+    private void writeType(ByteBuf buf, BType type) {
+        buf.writeInt(cp.addShapeCPEntry(type));
+    }
+
+    private void writeConstValue(ByteBuf buf, BLangConstantValue constValue) {
+        writeConstValue(buf, constValue.value, constValue.type);
+    }
+
+    private void writeConstValue(ByteBuf buf, Object value, BType type) {
+        switch (type.tag) {
+            case TypeTags.INT:
+            case TypeTags.SIGNED32_INT:
+            case TypeTags.SIGNED16_INT:
+            case TypeTags.SIGNED8_INT:
+            case TypeTags.UNSIGNED32_INT:
+            case TypeTags.UNSIGNED16_INT:
+            case TypeTags.UNSIGNED8_INT:
+                buf.writeInt(addIntCPEntry((Long) value));
+                break;
+            case TypeTags.BYTE:
+                int byteValue = ((Number) value).intValue();
+                buf.writeInt(addByteCPEntry(byteValue));
+                break;
+            case TypeTags.FLOAT:
+                // TODO:Remove the instanceof check by converting the float literal instance in Semantic analysis phase
+                double doubleVal = value instanceof String ? Double.parseDouble((String) value)
+                        : (Double) value;
+                buf.writeInt(addFloatCPEntry(doubleVal));
+                break;
+            case TypeTags.STRING:
+            case TypeTags.CHAR_STRING:
+            case TypeTags.DECIMAL:
+                buf.writeInt(addStringCPEntry((String) value));
+                break;
+            case TypeTags.BOOLEAN:
+                buf.writeBoolean((Boolean) value);
+                break;
+            case TypeTags.NIL:
+                break;
+            case TypeTags.RECORD:
+                Map<String, BLangConstantValue> mapConstVal = (Map<String, BLangConstantValue>) value;
+                buf.writeInt(mapConstVal.size());
+                mapConstVal.forEach((key, fieldValue) -> {
+                    buf.writeInt(addStringCPEntry(key));
+                    writeType(buf, fieldValue.type);
+                    writeConstValue(buf, fieldValue);
+                });
+                break;
+            case TypeTags.TUPLE:
+                BLangConstantValue[] tupleConstVal = (BLangConstantValue[]) value;
+                buf.writeInt(tupleConstVal.length);
+                for (BLangConstantValue memValue : tupleConstVal) {
+                    writeType(buf, memValue.type);
+                    writeConstValue(buf, memValue);
+                }
+                break;
+            case TypeTags.INTERSECTION:
+                BType effectiveType = ((BIntersectionType) type).effectiveType;
+                writeConstValue(buf, value, effectiveType);
+                break;
+            default:
+                // TODO support for other types
+                throw new UnsupportedOperationException(
+                        "finite type value is not supported for type: " + type);
+
+        }
+    }
```

Comment:
Aren't these methods duplicated?

Reply:
There are some minor differences (like `ConstValue` and `BLangConstantValue`). Will try to generalize the functions.